tools/gen-random-bout/gen-random-bout.cpp

```diff
@@ -22,116 +23,301 @@
 #define stat64 stat
 #endif
 
-// --sym-args 0 1 10 --sym-args 0 2 2 --sym-files 1 8 --sym-stdout
-static int getint(char *i) {
-  if(!i) {
+#define SMALL_BUFFER_SIZE 64 // To hold "arg<N>" string, temporary
+                             // filename, etc.
+
+#define MAX_FILE_SIZES 256
+
+static unsigned get_unsigned(char *i) {
+  if (!i) {
     fprintf(stderr, "ran out of arguments!\n");
     assert(i);
   }
-  return atoi(i);
+  return (unsigned)atoi(i);
 }
 
 #define MAX 64
-static void push_obj(KTest *b, const char *name, unsigned non_zero_bytes, 
-                     unsigned total_bytes) {
+static int push_random_obj(KTest *b, const char *name, unsigned non_zero_bytes,
+                           unsigned total_bytes) {
   KTestObject *o = &b->objects[b->numObjects++];
   assert(b->numObjects < MAX);
 
-  o->name = strdup(name);
+  if ((o->name = strdup(name)) == NULL) {
+    fprintf(stderr, "%s:%d: strdup() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
   o->numBytes = total_bytes;
-  o->bytes = (unsigned char *)malloc(o->numBytes);
+  if ((o->bytes = (unsigned char *)malloc(o->numBytes)) == NULL) {
+    fprintf(stderr, "%s:%d: malloc() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
 
   unsigned i;
-  for(i = 0; i < non_zero_bytes; i++)
-    o->bytes[i] = random();
+  for (i = 0; i < non_zero_bytes; i++) {
+    o->bytes[i] = random() % 255 + 1;
+  }
 
-  for(i = non_zero_bytes; i < total_bytes; i++)
+  for (i = non_zero_bytes; i < total_bytes; i++)
     o->bytes[i] = 0;
+
+  return 0;
 }
 
+static int push_obj(KTest *b, const char *name, unsigned total_bytes,
+                    unsigned char *content) {
+  KTestObject *o = &b->objects[b->numObjects++];
+  assert(b->numObjects < MAX);
+
+  if ((o->name = strdup(name)) == NULL) {
+    fprintf(stderr, "%s:%d: strdup() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
+  o->numBytes = total_bytes;
+  if ((o->bytes = (unsigned char *)malloc(total_bytes)) == NULL) {
+    fprintf(stderr, "%s:%d: malloc() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
+  memcpy(o->bytes, content, total_bytes);
+  return 0;
+}
 
-static void push_range(KTest *b, const char *name, unsigned value) {
+static int push_range(KTest *b, const char *name, unsigned value) {
   KTestObject *o = &b->objects[b->numObjects++];
   assert(b->numObjects < MAX);
 
-  o->name = strdup(name);
+  if ((o->name = strdup(name)) == NULL) {
+    fprintf(stderr, "%s:%d: strdup() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
   o->numBytes = 4;
-  o->bytes = (unsigned char *)malloc(o->numBytes);
+  if ((o->bytes = (unsigned char *)malloc(o->numBytes)) == NULL) {
+    fprintf(stderr, "%s:%d: malloc() failure\n", __FILE__, __LINE__);
+    return -1;
+  }
+
+  *(unsigned *)o->bytes = value;
+  return 0;
+}
+
+int create_stat(size_t size, struct stat64 *s) {
+  char filename[SMALL_BUFFER_SIZE];
```

Comment:
Instead of SMALL_BUFFER_SIZE you could use the length of the template string (just initialise and strdup). Also below: `mkstemp` returns a file descriptor, no need for `open`. You might add `fchmod` for the mode flags.